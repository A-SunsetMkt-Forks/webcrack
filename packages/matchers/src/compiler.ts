import type { NodePath, Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import {
  type ArraySchema,
  type CaptureMatcher,
  type Matcher,
  type NodeSchema,
  type NullableSchema,
  type OrMatcher,
  type Schema,
} from './types.js';

export type CompiledMatcher<T extends t.Node> = (
  input: T,
) => object | undefined;

interface Context {
  captures: string[];
}

/**
 * **Warning**: only compile a schema from trustworthy sources.
 * Otherwise, it could lead to code injection attacks.
 */
export function compile<T extends t.Node>(
  schema: NodeSchema<T>,
): CompiledMatcher<T> {
  const context: Context = { captures: [] };
  const checks = compileNode(schema, 'input', context, false);
  const code = `
const captures = { ${context.captures.map((v) => `${v}: undefined`).join(', ')} };
if (${checks ?? true}) { return captures; }`;

  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  return Function('input', code) as CompiledMatcher<T>;
}

/**
 * **Warning**: only compile a schema from trustworthy sources.
 * Otherwise, it could lead to code injection attacks.
 */
export function compileVisitor<T extends t.Node>(
  schema: NodeSchema<T>,
  cb: (path: NodePath<T>) => void,
  phase: 'enter' | 'exit' = 'enter',
): Visitor {
  const context: Context = { captures: [] };
  const checks = compileNode(schema, 'input', context, false);
  const code = `
const captures = { ${context.captures.map((v) => `${v}: undefined`).join(', ')} };
if (${checks ?? true}) { cb(captures); }`;

  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  const match = Function('input', code) as (
    input: T,
    cb: (path: NodePath<T>) => void,
  ) => void;

  return {
    [schema.type]: {
      [phase](path: NodePath<T>) {
        match(path.node, cb);
      },
    },
  };
}

type CompileFunction = (
  schema: never,
  input: string,
  ctx: Context,
) => string | null;

const compilers: Record<string, CompileFunction> = {
  nullable: compileNullable,
  tuple: compileTuple,
  arrayOf: compileArrayOf,
  or: compileOr,
  capture: compileCapture,
  fromCapture: () => null,
};
Object.keys(t.BUILDER_KEYS).forEach((type) => {
  compilers[type as t.Node['type']] = compileNode;
});

function compileSchema(
  schema: Schema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  if (schema === undefined) {
    return null;
  } else if (
    typeof schema === 'string' ||
    typeof schema === 'number' ||
    typeof schema === 'boolean' ||
    schema === null
  ) {
    return `${input} === ${JSON.stringify(schema)}`;
  } else if (Array.isArray(schema)) {
    return compileTuple(schema, input, ctx);
  } else {
    return compileMatcher(schema as Matcher<unknown>, input, ctx);
  }
}

function compileMatcher(
  matcher: Matcher<unknown>,
  input: string,
  ctx: Context,
): string | null {
  return compilers[matcher.type](matcher as never, input, ctx);
}

function compileTuple(
  schema: Schema<unknown>[],
  input: string,
  ctx: Context,
): string | null {
  const checks = [
    `${input}.length === ${schema.length}`,
    ...schema
      .map((m, i) => compileSchema(m, `${input}[${i}]`, ctx))
      .filter((c) => c !== null),
  ];
  return checks.join(' && ');
}

function compileArrayOf(
  schema: ArraySchema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  const checks = compileSchema(schema.schema, 'element', ctx);
  if (checks === null) return null;
  return `${input}.every(element => ${checks})`;
}

function compileNullable(
  schema: NullableSchema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  const checks = compileSchema(schema.schema, input, ctx);
  if (checks === null) return null;
  return `(${input} === null || ${checks})`;
}

function compileOr(schema: OrMatcher<unknown[]>, input: string, ctx: Context) {
  const checks = schema.schema
    .map((m) => compileSchema(m, input, ctx))
    .filter((c) => c !== null);
  if (checks.length === 0) return null;
  return `(${checks.join(' || ')})`;
}

function compileCapture(
  schema: CaptureMatcher<unknown>,
  input: string,
  ctx: Context,
) {
  ctx.captures.push(schema.name);
  const checks = compileSchema(schema.schema, input, ctx);
  if (checks === null) {
    return `(captures['${schema.name}'] = ${input}, true)`;
  } else {
    return `${checks} && (captures['${schema.name}'] = ${input}, true)`;
  }
}

function compileNode(
  schema: NodeSchema<t.Node>,
  input: string,
  ctx: Context,
  checkType = true,
) {
  const checks: string[] = [];
  if (checkType) {
    checks.push(`${input}.type === '${schema.type}'`);
  }

  const fields = t.NODE_FIELDS[schema.type];

  for (const key in fields) {
    const value = schema[key as never] as Schema<unknown>;
    const valueChecks = compileSchema(value, `${input}.${key}`, ctx);
    if (valueChecks === null) continue;

    const validator = fields[key];
    if (value !== null && validator.optional) {
      // TODO: only when all checks includes property accesses
      checks.push(`${input}.${key} != null`);
    }
    checks.push(valueChecks);
  }

  return checks.length ? checks.join(' && ') : null;
}
