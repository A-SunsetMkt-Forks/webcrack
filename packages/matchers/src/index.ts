export * from './compiler.js';
export * from './generated.js';
export * from './types.js';
