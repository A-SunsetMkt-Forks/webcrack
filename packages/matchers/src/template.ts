import { parse, parseExpression, type ParserOptions } from '@babel/parser';
import * as t from '@babel/types';
import { type NodeSchema } from './types.js';

export function expression(
  strings: TemplateStringsArray,
  ...schemas: NodeSchema<t.Node>[]
): NodeSchema<t.Expression> {
  return parseTemplate(
    strings,
    schemas,
    parseExpression,
  ) as NodeSchema<t.Expression>;
}

export function statement(
  strings: TemplateStringsArray,
  ...schemas: NodeSchema<t.Statement>[]
): NodeSchema<t.Statement> {
  return parseTemplate(
    strings,
    schemas,
    parseStatement,
  ) as NodeSchema<t.Statement>;
}

function parseStatement(input: string) {
  return parse(input, { allowReturnOutsideFunction: true }).program.body[0];
}

function parseTemplate(
  strings: TemplateStringsArray,
  schemas: NodeSchema<t.Node>[],
  parse: (input: string, options?: ParserOptions) => t.Node,
): NodeSchema<t.Node> {
  let schemaIndex = 0;
  const pattern = strings.reduce((acc, curr, i) => {
    acc += curr;

    if (schemas[i]) {
      acc += `$${schemaIndex++}`;
    }

    return acc;
  }, '');

  const ast = parse(pattern);
  let rootSchema: NodeSchema<t.Node> | undefined;
  t.traverse(ast, {
    enter(node, ancestors) {
      if (!t.isIdentifier(node)) return;
      if (!/^\$\d+$/.test(node.name)) return;
      const matcher = schemas[Number(node.name.slice(1))];
      const ancestor = ancestors.at(-1);
      if (ancestor) {
        const container = ancestor.node[ancestor.key as keyof t.Node];
        if (Array.isArray(container)) {
          container[ancestor.index!] = matcher as never;
        } else {
          ancestor.node[ancestor.key as keyof t.Node] = matcher as never;
        }
      } else {
        rootSchema = matcher;
      }
    },
  });

  return rootSchema ?? ast;
}
