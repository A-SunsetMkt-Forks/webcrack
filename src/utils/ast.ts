import generate from '@babel/generator';
import * as t from '@babel/types';

export function codePreview(node: t.Node): string {
  const { code } = generate(node, {
    minified: true,
    shouldPrintComment: () => false,
  });
  if (code.length > 100) {
    return code.slice(0, 70) + ' â€¦ ' + code.slice(-30);
  }
  return code;
}

export function getPropName(node: t.Node): string | undefined {
  if (t.isIdentifier(node)) {
    return node.name;
  }
  if (t.isStringLiteral(node)) {
    return node.value;
  }
  if (t.isNumericLiteral(node)) {
    return node.value.toString();
  }
}
