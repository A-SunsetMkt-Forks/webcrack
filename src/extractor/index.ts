import traverse, { Visitor, visitors } from '@babel/traverse';
import * as t from '@babel/types';
import * as m from '@codemod/matchers';
import debug from 'debug';
import { TransformState } from '../transforms';
import { unpackBrowserify } from './browserify';
import { Bundle } from './bundle';
import { unpackParcel } from './parcel';
import { unpackWebpack } from './webpack';

export function unpackBundle(
  ast: t.Node,
  mappings: Record<string, m.Matcher<unknown>> = {}
): Bundle | undefined {
  const options: { bundle: Bundle | undefined } = { bundle: undefined };
  const traverseOptions: Visitor<TransformState>[] = [
    unpackWebpack.visitor(options),
    unpackBrowserify.visitor(options),
    unpackParcel.visitor(options),
  ];
  const visitor = visitors.merge(traverseOptions);
  // @ts-expect-error regression from https://github.com/babel/babel/pull/15702
  visitor.noScope = traverseOptions.every(v => v.noScope);
  traverse(ast, visitor, undefined, { changes: 0 });
  if (options.bundle) {
    options.bundle.applyMappings(mappings);
    options.bundle.applyTransforms();
    debug('webcrack:unpack')('Bundle:', options.bundle.type);
  }
  return options.bundle;
}
