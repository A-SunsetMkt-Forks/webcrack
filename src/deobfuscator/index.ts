import debug from 'debug';
import {
  applyTransform,
  applyTransformAsync,
  applyTransforms,
  AsyncTransform,
} from '../transforms';
import mergeStrings from '../transforms/mergeStrings';
import { codePreview } from '../utils/ast';
import { findArrayRotator } from './arrayRotator';
import controlFlowObject from './controlFlowObject';
import controlFlowSwitch from './controlFlowSwitch';
import deadCode from './deadCode';
import { findDecoders } from './decoder';
import inlineDecodedStrings from './inlineDecodedStrings';
import inlineDecoderWrappers from './inlineDecoderWrappers';
import objectLiterals from './objectLiterals';
import { findStringArray } from './stringArray';
import { Sandbox, VMDecoder } from './vm';

// https://astexplorer.net/#/gist/b1018df4a8daebfcb1daf9d61fe17557/4ff9ad0e9c40b9616956f17f59a2d9888cd62a4f

export default {
  name: 'deobfuscate',
  tags: ['unsafe'],
  async run(ast, state, sandbox) {
    const logger = debug('webcrack:deobfuscate');
    if (!sandbox) return;

    const stringArray = findStringArray(ast);
    logger(`String Array: ${stringArray ? 'yes' : 'no'}`);
    if (!stringArray) return;
    logger(` - length: ${stringArray.length}`);
    logger(` - ${codePreview(stringArray.path.node)}`);

    const rotator = findArrayRotator(stringArray);
    logger(`String Array Rotate: ${rotator ? 'yes' : 'no'}`);
    if (rotator) {
      logger(` - ${codePreview(rotator.node)}`);
    }

    const decoders = findDecoders(stringArray);
    logger(`String Array Encodings: ${decoders.length}`);

    state.changes += applyTransform(ast, objectLiterals).changes;

    for (const decoder of decoders) {
      logger(` - ${codePreview(decoder.path.node)}`);

      state.changes += applyTransform(
        ast,
        inlineDecoderWrappers,
        decoder.path
      ).changes;
    }

    const vm = new VMDecoder(sandbox, stringArray, decoders, rotator);
    state.changes += (
      await applyTransformAsync(ast, inlineDecodedStrings, { vm })
    ).changes;

    stringArray.path.remove();
    rotator?.remove();
    decoders.forEach(decoder => decoder.path.remove());
    state.changes += 2 + decoders.length;

    state.changes += applyTransforms(
      ast,
      [mergeStrings, deadCode, controlFlowObject, controlFlowSwitch],
      'mergeStrings, deadCode, controlFlow'
    ).changes;
  },
} satisfies AsyncTransform<Sandbox>;
