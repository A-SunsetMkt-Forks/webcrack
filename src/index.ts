import generate from '@babel/generator';
import { parse } from '@babel/parser';
import * as m from '@codemod/matchers';
import debug from 'debug';
import { join } from 'node:path';
import deobfuscator from './deobfuscator';
import debugProtection from './deobfuscator/debugProtection';
import selfDefending from './deobfuscator/selfDefending';
import {
  Sandbox,
  createBrowserSandbox,
  createNodeSandbox,
} from './deobfuscator/vm';
import { unpackBundle } from './extractor';
import { Bundle } from './extractor/bundle';
import {
  applyTransform,
  applyTransformAsync,
  applyTransforms,
} from './transforms';
import blockStatement from './transforms/blockStatement';
import jsx from './transforms/jsx';
import sequence from './transforms/sequence';
import splitVariableDeclarations from './transforms/splitVariableDeclarations';
import unminify from './transforms/unminify';

export interface WebcrackResult {
  code: string;
  bundle: Bundle | undefined;
  /**
   * Save the deobufscated code and the extracted bundle to the given directory.
   * @param path Output directory
   */
  save(path: string): Promise<void>;
}

export interface Options {
  /**
   * Decompile react components to JSX.
   * @default true
   */
  jsx?: boolean;
  /**
   * Extract modules from the bundle.
   * @default true
   */
  unpack?: boolean;
  /**
   * Deobfuscate the code.
   * @default true
   */
  deobfuscate?: boolean;
  /**
   * Assigns paths to modules based on the given matchers.
   * This will also rewrite `require()` calls to use the new paths.
   *
   * @example
   * ```js
   * m => ({
   *   './utils/color.js': m.regExpLiteral('^#([0-9a-f]{3}){1,2}$')
   * })
   * ```
   */
  mappings?: (
    m: typeof import('@codemod/matchers')
  ) => Record<string, m.Matcher<unknown>>;
  /**
   * Function that executes a code expression and returns the result (typically from the obfuscator).
   */
  sandbox?: Sandbox;
}

function mergeOptions(options: Options): asserts options is Required<Options> {
  const mergedOptions: Required<Options> = {
    jsx: true,
    unpack: true,
    deobfuscate: true,
    mappings: () => ({}),
    sandbox: process.env.browser ? createBrowserSandbox() : createNodeSandbox(),
    ...options,
  };
  Object.assign(options, mergedOptions);
}

export async function webcrack(
  code: string,
  options: Options = {}
): Promise<WebcrackResult> {
  mergeOptions(options);

  if (process.env.browser) {
    debug.enable('webcrack:*');
  }

  const ast = parse(code, {
    sourceType: 'unambiguous',
    allowReturnOutsideFunction: true,
    plugins: ['jsx'],
  });

  applyTransforms(
    ast,
    [blockStatement, sequence, splitVariableDeclarations],
    'prepare'
  );

  if (options.deobfuscate)
    await applyTransformAsync(ast, deobfuscator, options.sandbox);

  applyTransform(ast, unminify);

  // TODO: Also merge unminify visitor (breaks selfDefending/debugProtection atm)
  applyTransforms(
    ast,
    [
      // Have to run this after unminify to properly detect it
      options.deobfuscate ? [selfDefending, debugProtection] : [],
      options.jsx ? [jsx] : [],
    ].flat()
  );

  // Unpacking modifies the same AST and may result in imports not at top level
  // so the code has to be generated before
  const outputCode = generate(ast).code;

  const bundle = options.unpack
    ? unpackBundle(ast, options.mappings(m))
    : undefined;

  return {
    code: outputCode,
    bundle,
    async save(path) {
      if (process.env.browser) {
        throw new Error('Not implemented.');
      } else {
        const { mkdir, writeFile } = await import('node:fs/promises');
        await mkdir(path, { recursive: true });
        await writeFile(join(path, 'deobfuscated.js'), outputCode, 'utf8');
        await bundle?.save(path);
      }
    },
  };
}
