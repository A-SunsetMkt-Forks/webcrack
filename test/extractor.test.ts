import assert from 'assert';
import { readFile } from 'fs/promises';
import { tmpdir } from 'os';
import { join } from 'path';
import { describe, expect, test } from 'vitest';
import { webcrack } from '../src/index';
import { relativePath, resolveDependencyTree } from '../src/utils/path';

// Test samples
test.each([
  'webpack.js',
  'webpack-object.js',
  'webpack-esm.js',
  'webpack-var-injection.js',
  'webpack-jsonp-chunk.js',
  'webpack-0.11.x.js',
  'webpack5-object.js',
  'webpack5-esm.js',
  'browserify.js',
  'browserify-2.js',
])('extract %s', async filename => {
  const { bundle } = await webcrack(
    await readFile(join('test', 'samples', filename), 'utf8')
  );
  expect(bundle).toMatchSnapshot();
});

test('detect top-level bundle first', async () => {
  const { bundle } = await webcrack(
    await readFile(
      join('test', 'samples', 'browserify-webpack-nested.js'),
      'utf8'
    )
  );
  assert(bundle);
  expect(bundle.type).toBe('browserify');
});

describe('extractor', () => {
  test('path mapping', async () => {
    const { bundle } = await webcrack(
      await readFile('./test/samples/webpack.js', 'utf8'),
      {
        mappings: m => ({
          './utils/color.js': m.stringLiteral('#FBC02D'),
          package: m.numericLiteral(4),
        }),
      }
    );
    expect(bundle).toBeDefined();
    assert(bundle);
    expect(bundle).toMatchSnapshot();
  });
});

test('prevent path traversal', async () => {
  const code = await readFile('test/samples/webpack-path-traversal.js', 'utf8');
  const result = await webcrack(code);
  const dir = join(tmpdir(), 'path-traversal-test');
  await expect(result.save(dir)).rejects.toThrow('path traversal');
});

describe('paths', () => {
  test('relative paths', () => {
    expect(relativePath('./a.js', './x/y.js')).toBe('./x/y.js');
    expect(relativePath('./x/y.js', './a.js')).toBe('../a.js');
    expect(relativePath('./a.js', 'node_modules/lib')).toBe('lib');
  });

  test('resolve browserify paths', () => {
    const dependencies = {
      0: { 1: './a.js', 4: 'lib' },
      1: { 2: '../bar/b.js' },
      2: { 3: '../../c.js' },
      3: {},
      4: {},
    };
    expect(resolveDependencyTree(dependencies, '0')).toEqual({
      0: 'tmp0/tmp1/index.js',
      1: 'tmp0/tmp1/a.js',
      2: 'tmp0/bar/b.js',
      3: 'c.js',
      4: 'node_modules/lib/index.js',
    });
  });

  test('resolve browserify paths 2', () => {
    const dependencies = {
      1: {},
      2: { 5: './v1', 6: './v4' },
      3: {},
      4: {},
      5: { 3: './lib/bytesToUuid', 4: './lib/rng' },
      6: { 3: './lib/bytesToUuid', 4: './lib/rng' },
      7: { 1: 'number', 2: 'uuid' },
    };
    expect(resolveDependencyTree(dependencies, '7')).toEqual({
      1: 'node_modules/number/index.js',
      2: 'node_modules/uuid/index.js',
      3: 'node_modules/uuid/lib/bytesToUuid.js',
      4: 'node_modules/uuid/lib/rng.js',
      5: 'node_modules/uuid/v1.js',
      6: 'node_modules/uuid/v4.js',
      7: 'index.js',
    });
  });
});
