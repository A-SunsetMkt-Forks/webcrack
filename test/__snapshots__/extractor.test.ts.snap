// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract browserify.js 1`] = `
BrowserifyBundle {
  "entryId": "2",
  "modules": Map {
    "1" => BrowserifyModule {
      "ast": function add(a, b) {
  return a + b;
}
module.exports = add;,
      "dependencies": {},
      "id": "1",
      "isEntry": false,
      "path": "add.js",
    },
    "2" => BrowserifyModule {
      "ast": var sum = require("./sum");
var numbers = [1, 2, 3];
var result = sum(numbers);
var outputElement = document.getElementById("output");
outputElement.innerHTML = result;,
      "dependencies": {
        "4": "./sum",
      },
      "id": "2",
      "isEntry": true,
      "path": "index.js",
    },
    "3" => BrowserifyModule {
      "ast": function reduce(list, iteratee, memo) {
  list.forEach(function (item) {
    memo = iteratee(item, memo);
  });
  return memo;
}
module.exports = reduce;,
      "dependencies": {},
      "id": "3",
      "isEntry": false,
      "path": "reduce.js",
    },
    "4" => BrowserifyModule {
      "ast": var reduce = require("./reduce");
var add = require("./add");
function sum(list) {
  return reduce(list, add, 0);
}
module.exports = sum;,
      "dependencies": {
        "1": "./add",
        "3": "./reduce",
      },
      "id": "4",
      "isEntry": false,
      "path": "sum.js",
    },
  },
  "type": "browserify",
}
`;

exports[`extract browserify-2.js 1`] = `
BrowserifyBundle {
  "entryId": "2",
  "modules": Map {
    "1" => BrowserifyModule {
      "ast": module.exports = 1;,
      "dependencies": {},
      "id": "1",
      "isEntry": false,
      "path": "lib.js",
    },
    "2" => BrowserifyModule {
      "ast": const vscode = require("vscode");
const lib = require("./lib");
console.log(lib);,
      "dependencies": {
        "1": "./lib",
      },
      "id": "2",
      "isEntry": true,
      "path": "index.js",
    },
  },
  "type": "browserify",
}
`;

exports[`extract webpack.js 1`] = `
WebpackBundle {
  "entryId": "2",
  "modules": Map {
    "1" => WebpackModule {
      "ast": const a = require("./3.js");,
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": const a = require("./1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;,
      "id": "2",
      "isEntry": true,
      "path": "./index.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = 4;,
      "id": "3",
      "isEntry": false,
      "path": "./3.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack-0.11.x.js 1`] = `
WebpackBundle {
  "entryId": "0",
  "modules": Map {
    "0" => WebpackModule {
      "ast": require("./1.js");
var template = require("./4.js");
document.write(template({
  hello: "World!"
}));,
      "id": "0",
      "isEntry": true,
      "path": "./index.js",
    },
    "1" => WebpackModule {
      "ast": // style-loader: Adds some css to the DOM by adding a <style> tag
var dispose = require("./2.js")
// The css code:
(require("./3.js"));
if (false) {
  module.hot.accept();
  module.hot.dispose(dispose);
},
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": /*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
module.exports = function (cssCode) {
  var styleElement = document.createElement("style");
  styleElement.type = "text/css";
  if (styleElement.styleSheet) {
    styleElement.styleSheet.cssText = cssCode;
  } else {
    styleElement.appendChild(document.createTextNode(cssCode));
  }
  var head = document.getElementsByTagName("head")[0];
  head.appendChild(styleElement);
  return function () {
    head.removeChild(styleElement);
  };
};,
      "id": "2",
      "isEntry": false,
      "path": "./2.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = "body {\\n\\tbackground: #333;\\n\\tcolor: #EEE;\\n}";,
      "id": "3",
      "isEntry": false,
      "path": "./3.js",
    },
    "4" => WebpackModule {
      "ast": var jade = require("./5.js");
module.exports = function anonymous(locals, attrs, escape, rethrow, merge) {
  attrs = attrs || jade.attrs;
  escape = escape || jade.escape;
  rethrow = rethrow || jade.rethrow;
  merge = merge || jade.merge;
  var buf = [];
  with (locals || {}) {
    var interp;
    buf.push("<h1>Hello " + escape((interp = hello) == null ? "" : interp) + "</h1>");
  }
  return buf.join("");
};,
      "id": "4",
      "isEntry": false,
      "path": "./4.js",
    },
    "5" => WebpackModule {
      "ast": /*!
 * Jade - runtime
 * Copyright(c) 2010 TJ Holowaychuk <tj@vision-media.ca>
 * MIT Licensed
 */

/**
 * Lame Array.isArray() polyfill for now.
 */

if (!Array.isArray) {
  Array.isArray = function (arr) {
    return Object.prototype.toString.call(arr) == "[object Array]";
  };
}

/**
 * Lame Object.keys() polyfill for now.
 */

if (!Object.keys) {
  Object.keys = function (obj) {
    var arr = [];
    for (var key in obj) {
      if (obj.hasOwnProperty(key)) {
        arr.push(key);
      }
    }
    return arr;
  };
}

/**
 * Merge two attribute objects giving precedence
 * to values in object \`b\`. Classes are special-cased
 * allowing for arrays and merging/joining appropriately
 * resulting in a string.
 *
 * @param {Object} a
 * @param {Object} b
 * @return {Object} a
 * @api private
 */

exports.merge = function merge(a, b) {
  var ac = a.class;
  var bc = b.class;
  if (ac || bc) {
    ac = ac || [];
    bc = bc || [];
    if (!Array.isArray(ac)) {
      ac = [ac];
    }
    if (!Array.isArray(bc)) {
      bc = [bc];
    }
    ac = ac.filter(nulls);
    bc = bc.filter(nulls);
    a.class = ac.concat(bc).join(" ");
  }
  for (var key in b) {
    if (key != "class") {
      a[key] = b[key];
    }
  }
  return a;
};

/**
 * Filter null \`val\`s.
 *
 * @param {Mixed} val
 * @return {Mixed}
 * @api private
 */

function nulls(val) {
  return val != null;
}

/**
 * Render the given attributes object.
 *
 * @param {Object} obj
 * @param {Object} escaped
 * @return {String}
 * @api private
 */

exports.attrs = function attrs(obj, escaped) {
  var buf = [];
  var terse = obj.terse;
  delete obj.terse;
  var keys = Object.keys(obj);
  var len = keys.length;
  if (len) {
    buf.push("");
    for (var i = 0; i < len; ++i) {
      var key = keys[i];
      var val = obj[key];
      if (typeof val == "boolean" || val == null) {
        if (val) {
          if (terse) {
            buf.push(key);
          } else {
            buf.push(key + "=\\"" + key + "\\"");
          }
        }
      } else if (key.indexOf("data") == 0 && typeof val != "string") {
        buf.push(key + "='" + JSON.stringify(val) + "'");
      } else if (key == "class" && Array.isArray(val)) {
        buf.push(key + "=\\"" + exports.escape(val.join(" ")) + "\\"");
      } else if (escaped && escaped[key]) {
        buf.push(key + "=\\"" + exports.escape(val) + "\\"");
      } else {
        buf.push(key + "=\\"" + val + "\\"");
      }
    }
  }
  return buf.join(" ");
};

/**
 * Escape the given string of \`html\`.
 *
 * @param {String} html
 * @return {String}
 * @api private
 */

exports.escape = function escape(html) {
  return String(html).replace(/&(?!(\\w+|\\#\\d+);)/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
};

/**
 * Re-throw the given \`err\` in context to the
 * the jade in \`filename\` at the given \`lineno\`.
 *
 * @param {Error} err
 * @param {String} filename
 * @param {String} lineno
 * @api private
 */

exports.rethrow = function rethrow(err, filename, lineno) {
  if (!filename) {
    throw err;
  }
  var context = 3;
  var str = require("./6.js").readFileSync(filename, "utf8");
  var lines = str.split("\\n");
  var start = Math.max(lineno - context, 0);
  var end = Math.min(lines.length, lineno + context); // Error context
  var context = lines.slice(start, end).map(function (line, i) {
    var curr = i + start + 1;
    return (curr == lineno ? "  > " : "    ") + curr + "| " + line;
  }).join("\\n");

  // Alter exception message
  err.path = filename;
  err.message = (filename || "Jade") + ":" + lineno + "\\n" + context + "\\n\\n" + err.message;
  throw err;
};,
      "id": "5",
      "isEntry": false,
      "path": "./5.js",
    },
    "6" => WebpackModule {
      "ast": var files = {};
exports.setFile = function (filename, content) {
  files[filename] = content;
};
exports.readFileSync = function (filename) {
  return files[filename] || "";
};,
      "id": "6",
      "isEntry": false,
      "path": "./6.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack-esm.js 1`] = `
WebpackBundle {
  "entryId": "0",
  "modules": Map {
    "0" => WebpackModule {
      "ast": const a = require("./1.js");
const b = require("./2.js");
const c = require("./3.js");
const bDefault = b.default;
const cDefault = c;
const dDefault = b.default;
const eDefault = c;
console.log(a.counter);
console.log(bDefault.VERSION);
console.log(cDefault.VERSION);,
      "id": "0",
      "isEntry": true,
      "path": "./index.js",
    },
    "1" => WebpackModule {
      "ast": export class counter {}
let _counter = 1;
counter = 2;,
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": const x = {};
export default 1;,
      "id": "2",
      "isEntry": false,
      "path": "./2.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = {
  VERSION: 1
};,
      "id": "3",
      "isEntry": false,
      "path": "./3.js",
    },
    "4" => WebpackModule {
      "ast": export default class f {},
      "id": "4",
      "isEntry": false,
      "path": "./4.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack-jsonp-chunk.js 1`] = `
WebpackBundle {
  "entryId": "",
  "modules": Map {
    "2wwy" => WebpackModule {
      "ast": const a = require("./E8gZ.js");
const b = require( /*webcrack:missing*/"./w6GO.js");,
      "id": "2wwy",
      "isEntry": false,
      "path": "./2wwy.js",
    },
    "E8gZ" => WebpackModule {
      "ast": module.exports = 4;,
      "id": "E8gZ",
      "isEntry": false,
      "path": "./E8gZ.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack-object.js 1`] = `
WebpackBundle {
  "entryId": "386",
  "modules": Map {
    "386" => WebpackModule {
      "ast": const r = require("./387.js").default;,
      "id": "386",
      "isEntry": true,
      "path": "./index.js",
    },
    "387" => WebpackModule {
      "ast": module.exports = "test";,
      "id": "387",
      "isEntry": false,
      "path": "./387.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack-var-injection.js 1`] = `
WebpackBundle {
  "entryId": "0",
  "modules": Map {
    "0" => WebpackModule {
      "ast": import * as m from "./1.js";
import * as n from "./2.js";
console.log(m, n);,
      "id": "0",
      "isEntry": true,
      "path": "./index.js",
    },
    "1" => WebpackModule {
      "ast": module.exports = 1;,
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": module.exports = 2;,
      "id": "2",
      "isEntry": false,
      "path": "./2.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack5-esm.js 1`] = `
WebpackBundle {
  "entryId": "2",
  "modules": Map {
    "2" => WebpackModule {
      "ast": const lib = require("./3.js");
console.log(lib);,
      "id": "2",
      "isEntry": true,
      "path": "./index.js",
    },
    "3" => WebpackModule {
      "ast": import * as a from "./4.js";
const obj = {
  version: "2.0.0"
};
export let version = obj.version;
export default a.foo;,
      "id": "3",
      "isEntry": false,
      "path": "./3.js",
    },
    "4" => WebpackModule {
      "ast": export const foo = 2;
export var obj = {
  Console: bar
};
function bar() {},
      "id": "4",
      "isEntry": false,
      "path": "./4.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extract webpack5-object.js 1`] = `
WebpackBundle {
  "entryId": "2",
  "modules": Map {
    "2" => WebpackModule {
      "ast": const lib = require("./3.js");
console.log(lib);
const _0x8da276 = require(require.ab + "build/Release/spdlog.node");,
      "id": "2",
      "isEntry": true,
      "path": "./index.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = "foo";,
      "id": "3",
      "isEntry": false,
      "path": "./3.js",
    },
  },
  "type": "webpack",
}
`;

exports[`extractor > path mapping 1`] = `
WebpackBundle {
  "entryId": "2",
  "modules": Map {
    "1" => WebpackModule {
      "ast": const a = require("package");,
      "id": "1",
      "isEntry": false,
      "path": "./1.js",
    },
    "2" => WebpackModule {
      "ast": const a = require("../1.js");
const _module = 1;
module.exports.color = "#FBC02D";
{
  const _module2 = 2;
  console.log(_module2);
  console.log(module);
}
exports.a = 3;,
      "id": "2",
      "isEntry": true,
      "path": "./utils/color.js",
    },
    "3" => WebpackModule {
      "ast": module.exports = 4;,
      "id": "3",
      "isEntry": false,
      "path": "node_modules/package",
    },
  },
  "type": "webpack",
}
`;
